### ISSUE TITLE
Speculative reads based on read latency profile may cease to be effective temporarily

### ISSUE THREAD

**Issue by tgrabiec** [scylladb/scylla#3746]
By default, speculative reads are configured to trigger for replica reads (before reaching CL) taking longer than the 99th percentile according to the read latency histogram (called the histogram from now on). Even though the histogram is used as a baseline for replica reads, it is updated from the duration of the whole coordinator read (when CL is met and reconciled). This duration is sometimes significantly longer than the duration of replica reads involved in reaching the CL. This may cause the 99th percentile of the histogram to drift up towards the coordinator-side latencies, which in some cases effectively disables speculation for replica reads.

One case when it happens is when reads are speculating, and the read is satisfied from the speculated request. Let's assume RF=3, CL=QUORUM, and that in the past all 3 replicas responded in 10 ms. Now suppose 1 of them gets slower and starts to respond in 100ms, but is not fully utilized, so throughput doesn't suffer. Initially, the 99th percentile is 10ms, and 2/3 of reads will speculate. The reads which speculate will complete in 20ms, because we issue the speculating read after reaching the 99th percentile, so after 10ms. That will change the 99th percentile to be now 20ms. The reads are still speculating, but speculation will now happen 20ms into the read, and will now complete in 30ms. In the next cycle it will complete in 40ms, and so on until we reach 100ms and speculation stops. This can quickly make speculation ineffective in covering the slowness. Obsereved in https://github.com/scylladb/scylla/issues/3628#issuecomment-415441770.

In this particular scenario, it would be better to keep speculating at the old threshold, so that the client latency remains low until the high latency period goes away. This would be achieved by making sure we update the histogram from the duration of replica reads which contributed to the CL, rather than the duration of a coordinator read. This way the histogram wouldn't drift up.

Another potential case where we could see this effect is when a lot of reads need reconciliation.

Fixing this may badly interact with our lack of backpressure for background reads. If slowness lasts long and is also associated with lower throughput, it may cause lots of background reads to accumulate with time. So before fixing this, we should have appropriate backpressure mechanisms in place which prevent the overload.

**Comment by kbr-** [comment]
@tgrabiec do you know if we already have the aforementioned "appropriate backpressure mechanisms"? @psarna is this what you're currently working on?

I wanted to fix #7342 by making the metric more stable / have larger momentum (see my last comment: https://github.com/scylladb/scylla/issues/7342#issuecomment-761003384) but after performing more experiments it turns out the "fix" doesn't always work, and second - it won't fix the issue that you described here, i.e. replica slowness (it only addresses replica crash) - if the replica is slow for a sufficiently long time, the metric would degrade eventually anyway.

I really like the solution you and @piodul  propose - measuring replica round-trip latencies, not the entire coordinator read - but the background read accumulation is a problem. Your prediction is spot on - the experiments with node crash showed it clearly in metrics, and they accumulate FAST; then they slowly drop most probably due to the the latency increase caused by speculative retries, so this latency increase kind of acts like the backpressure mechanism today.
![image](https://user-images.githubusercontent.com/6516501/104912692-e22fa980-598c-11eb-8074-49cb1ce39650.png)

**Comment by tgrabiec** [comment]
@kbr- I don't think we added any backpressure to the read path.

**Comment by vladzcloudius** [comment]
@tgrabiec @eliransin @kbr- Guys, keep in mind that there is a much better/safer way to configure speculative reads: latency based  (as opposed to a latency percentile based).

Namely you can configure the speculative retry to kick in only after it takes X time for a request to achieve the consistency.

While it is not going to prevent the avalanche effect describe in this GH issue (and in a few others) - it's making using the speculative reads much easier. 

For instance, if your read timeouts are configured to 5s you can set the speculative_retry to 4s.

**Comment by avikivity** [comment]
Performance, not backporting.
